File: 104.maximum-depth-of-binary-tree.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        return self.dfs(root, 0)+1
    def dfs(self, root: Optional[TreeNode], answer) -> int:
        if root is None:
            answer -= 1
            return answer
        answer += 1
        return max(self.dfs(root.left, answer), self.dfs(root.right, answer))


# @leet end

File: 104.maximum-depth-of-binary-tree.java
// @leet start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  public int maxDepth(TreeNode root) { return dfs(root, 1); }
  public int dfs(TreeNode root, int answer) {
    if (root == null) {
      return answer - 1;
    }
    answer++;
    return Math.max(dfs(root.left, answer), dfs(root.right, answer));
  }
}
// @leet end
